### PR TITLE
Update default StokesWeights operator

### PR DESCRIPTION
**Pull request by tskisner**
When generating the elements of the pointing matrix.  The mathematical basis for these changes can be found in this document:

[HWP_algebra.pdf](https://github.com/hpc4cmb/toast/files/11494763/HWP_algebra.pdf)

One important note is that after this PR is merged, all instrument models for workflows that use a HWP will need to ensure that a focalplane table column exists for the detector gamma angle.

* Update sign convention to match accompanying document (to be added to documentation) given our recent formalization of the focalplane coordinate frame.

* When using a HWP, require per-detector offsets in the focalplane table describing the fixed angle from focalplane frame X-axis to detector frame X-axis.

* Correctly handle the cases of fixed HWP angle versus no HWP.

* Add unit tests that cover several permutations of I/Q/U sky pixel values, fixed HWP angles, and detector orientations.

### CODE REVIEW

**Comment by tskisner** [comment]
Ok, this is finally ready for review.

File: src/toast/ops/stokes_weights/stokes_weights.py [hunk 1]
```diff
@@ -149,6 +152,10 @@ def _exec(self, data, detectors=None, use_accel=None, **kwargs):
         if self.detector_pointing is None:
             raise RuntimeError("The detector_pointing trait must be set")
 
+        if (self.mode == "IQU") and self.hwp_angle is not None:
```
Comment (keskitalo) [inline src/toast/ops/stokes_weights/stokes_weights.py:155]: There could be other pointing modes that require `gamma`. Namely, any polarized mode, including `QU` and `IQUSS`.  How about checking  `if "QU" in self.mode` instead?

**Review by keskitalo — COMMENTED** [review]
Looks great! Only minor observations. Very happy to see a dedicated stokes weights unit test.

File: src/toast/instrument_sim.py [hunk 2]
```diff
@@ -577,13 +634,15 @@ def fake_hexagon_focalplane(
 
     nominal_freq = str(int(bandcenter.to_value(u.GHz)))
     det_names = [f"{x}-{nominal_freq}" for x in det_data.keys()]
+    det_gamma = [dgamma[x] for x in det_data.keys()]
```
Comment (giuspugl) [inline src/toast/instrument_sim.py:637]: we need to propagate this change to all  Instrument Models (IMO)  that use toast3. What happens if an IMO doesn't encode that ? toast should raise an error, shouldn't it ?

File: src/toast/ops/stokes_weights/stokes_weights.py [hunk 3]
```diff
@@ -24,33 +25,31 @@ class StokesWeights(Operator):
...
+    measured from the local meridian.  The omega value can be described in terms of
+    alpha, a fixed per-detector offset gamma, and the time varying HWP angle measured
+    from the focalplane coordinate frame X-axis:
 
     .. math::
-        v_m = vector parallel to local meridian
-        v_o = orientation vector (transformed x-axis)
-        v_d = direction vector (transformed z-axis)
-        a = atan2((v_m X v_o) \\dot v_d, v_m \\dot v_o)
+        \\omega = \\alpha + {\\gamma}_{H}(t) - {\\gamma}_{D}
```
Comment (giuspugl) [inline src/toast/ops/stokes_weights/stokes_weights.py:50]: to be more explicit, how about : 
`\\omega = \\alpha + {\\gamma}_{HWP}(t) - {\\gamma}_{DET}` 
?

File: src/toast/ops/stokes_weights/kernels_jax.py [hunk 4]
```diff
@@ -13,39 +13,57 @@
...
-    # computes detang
-    detang = jnp.arctan2(by, bx)
-    detang = detang + 2.0 * hwpang
-    detang = 2.0 * detang
-
-    # puts values into weights
-    eta = (1.0 - eps) / (1.0 + eps)
-    weights = jnp.array([cal, jnp.cos(detang) * eta * cal, jnp.sin(detang) * eta * cal])
+    if use_hwp:
+        ang = 2.0 * (alpha + 2.0 * (hwpang - gamma))
```
Comment (giuspugl) [inline src/toast/ops/stokes_weights/kernels_jax.py:61]: Ok here we multiply by 4 `hwpang`

File: src/toast/ops/conviqt.py [hunk 5]
```diff
@@ -805,30 +851,35 @@ def _exec(self, data, detectors=None, **kwargs):
...
-            pnt = self.get_buffer(theta, phi, psi, det, verbose)
+            pnt = self.get_buffer(theta, phi, psi_beam, det, verbose)
 
             convolved_data = self.convolve(
                 skyT, beam_T, detector, pnt, det, verbose, pol=False
             )
 
             if self.pol:
                 del (pnt,)
+                angle_arg = 4.0 * hwp_angle
```
Comment (giuspugl) [inline src/toast/ops/conviqt.py:867]: We shouldn't multiply here by 4 , should we ? see above https://github.com/hpc4cmb/toast/pull/663/files#r1221116956

File: src/toast/ops/stokes_weights/kernels_jax.py [hunk 6]
```diff
@@ -13,39 +13,57 @@
 from ...utils import Logger
 
 
-def stokes_weights_IQU_inner(eps, cal, pin, hwpang):
+def stokes_weights_IQU_inner(eps, cal, gamma, pin, hwpang, use_hwp, IAU):
```
[the remark this answers is not in the document]
Comment (tskisner) [reply src/toast/ops/stokes_weights/kernels_jax.py]: The jax kernel was just updated, and it now works locally.  Unfortunately no automated testing yet on GPUs.

File: src/toast/ops/stokes_weights/stokes_weights.py (hunk 1 above)
Comment (tskisner) [reply src/toast/ops/stokes_weights/stokes_weights.py]: Changed.

File: src/toast/instrument_sim.py (hunk 2 above)
Comment (tskisner) [reply src/toast/instrument_sim.py:637]: There is an exception raised if the `hwp_angle` is not `None` in the StokesWeights operator and the focalplane is missing the `gamma` key.

File: src/toast/ops/stokes_weights/stokes_weights.py (hunk 3 above)
Comment (tskisner) [reply src/toast/ops/stokes_weights/stokes_weights.py]: Fixed.

File: src/toast/ops/stokes_weights/kernels_jax.py [hunk 7]
```diff
@@ -116,17 +191,18 @@ def stokes_weights_IQU_interval(
         quats, (quat_index, intervals, ALL)
     )  # quats[quat_index,intervals,:]
 
-    # insures hwp is a non empty array
+    IAU_sign = 1.0 - 2.0 * IAU
```
Comment (nestordemeure) [inline src/toast/ops/stokes_weights/kernels_jax.py:194]: This could use the [sign function](https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.sign.html).

File: src/toast/ops/stokes_weights/kernels_jax.py [hunk 8]
```diff
@@ -81,7 +152,9 @@ def stokes_weights_IQU_interval(
     weights,
     hwp,
     epsilon,
+    gamma,
     cal,
+    IAU,
```
Comment (nestordemeure) [inline src/toast/ops/stokes_weights/kernels_jax.py:157]: IAU could be added to the static arugments in the jit

File: src/toast/ops/stokes_weights/kernels_numpy.py [hunk 9]
```diff
@@ -10,10 +10,26 @@
...
+    epsilon,
+    gamma,
+    cal,
+    IAU,
+    use_accel,
 ):
     if hwp is not None and len(hwp) == 0:
         hwp = None
+
+    if IAU:
```
Comment (nestordemeure) [inline src/toast/ops/stokes_weights/kernels_numpy.py:28]: This could use the numpy sign function

File: src/toast/ops/stokes_weights/kernels_jax.py (hunk 7 above)
Comment (nestordemeure) [reply src/toast/ops/stokes_weights/kernels_jax.py:194]: (or a one line test if IAU is static, something like: `IAU_sign = 1 if IAU else -1`)
